daemon/state.rs

```diff
@@ -0,0 +1,99 @@
+use atomicwrites::{AtomicFile, OverwriteBehavior};
+use serde::{Deserialize, Serialize};
+use std::io::Write;
+use std::path::Path;
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct State {
+    pub total_jobs_completed: u64,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        Self {
+            total_jobs_completed: 0,
+        }
+    }
+}
+
+pub fn load(state_file: &Path) -> State {
+    log::debug!("Loading initial state from {}", state_file.display());
+    match std::fs::read_to_string(state_file) {
+        Err(err) => {
+            if err.kind() != std::io::ErrorKind::NotFound {
+                log::warn!(
+                    "Cannot read initial state from {}: {}",
+                    state_file.display(),
+                    err
+                );
+            }
+            return State::default();
+        }
+        Ok(data) => match serde_json::from_str::<State>(&data) {
+            Err(err) => {
+                log::warn!(
+                    "Cannot parse initial state from {}: {}",
+                    state_file.display(),
+                    err
+                );
+                return State::default();
+            }
+            Ok(state) => {
+                log::debug!("Initial state: {:?}", state);
+                return state;
+            }
+        },
+    }
+}
+
+pub fn store(state_file: &Path, state: &State) {
+    if let Some(parent) = state_file.parent() {
+        if let Err(err) = std::fs::create_dir_all(&parent) {
+            log::warn!(
+                "Cannot create state directory {}: {}",
+                parent.display(),
+                err
+            );
+            return;
+        }
+    }
+
+    let payload = match serde_json::to_string_pretty(&state) {
+        Err(err) => {
+            log::warn!("Cannot serialize state: {}", err);
+            return;
```

Comment:
should we return an error here instead, so that the consumer can handle it (or it can make the process crash)?

Reply:
This serialisation call should not fail unless we make a programmer error. Quoting from `serde_json::to_string_pretty` docs:

```rs
/// Serialization can fail if `T`'s implementation of `Serialize` decides to
/// fail, or if `T` contains a map with non-string keys.
```

If our State becomes impossible to serialise to JSON, then we are pretty much doomed. If we crash the process, then Station should restart it. This will lead to a State that cannot be serialised, so the process crashes again, and Station tries to restart it again. And so on.

By ignoring the error, we can keep running jobs and earning rewards, although the job counter will show incorrect value.

It would be great if `zinniad` could report to us (via Sentry?) when we encounter this problem, but I feel that's beyond the scope of this pull request.

WDYT?

Reply:
> If our State becomes impossible to serialise to JSON, then we are pretty much doomed. If we crash the process, then Station should restart it. This will lead to a State that cannot be serialised, so the process crashes again, and Station tries to restart it again. And so on.

To me that's better than running modules but being fundamentally unable to deal with their results. Station should give up after a while and report there is a series problem. It already will do that, using Sentry.

Reply:
Fair enough. Should I crash the process on all errors, including I/O errors when writing the state file, or only when we cannot serialise to JSON?

Reply:
I think only on fatal errors that you can't recover from. Failing to write a file once can be recoverable, being unable to serialize state not

Reply:
In the call, we agreed to crash on all errors for now. I made the changes in 6055e99. I have to say I like the new code much more. Thanks for the nudge to improve this part!